File: services/forge-control/src/main/kotlin/forge/control/logging/JsonLog.kt
package forge.control.logging

import forge.control.telemetry.LoggingContext
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

private val LEVEL_RANK = mapOf(
    "debug" to 10,
    "info" to 20,
    "warn" to 30,
    "error" to 40,
)

private val TIMESTAMP_FMT = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'")
    .withZone(ZoneOffset.UTC)

/** Structured JSON logging to stdout (timestamp, level, service, message). */
class JsonLog(
    private val service: String,
    level: String,
) {
    private val min = LEVEL_RANK[level.lowercase()] ?: 20

    fun debug(message: String, vararg fields: Pair<String, Any?>) = emit("debug", message, fields)
    fun info(message: String, vararg fields: Pair<String, Any?>) = emit("info", message, fields)
    fun warn(message: String, vararg fields: Pair<String, Any?>) = emit("warn", message, fields)
    fun error(message: String, vararg fields: Pair<String, Any?>) = emit("error", message, fields)

    private fun emit(level: String, message: String, fields: Array<out Pair<String, Any?>>) {
        if ((LEVEL_RANK[level] ?: 20) < min) return
        val payload = buildJsonObject {
            put("timestamp", JsonPrimitive(TIMESTAMP_FMT.format(Instant.now())))
            put("level", JsonPrimitive(level))
            put("service", JsonPrimitive(service))
            put("message", JsonPrimitive(message))
            val correlation = LoggingContext.current()
            put("requestId", JsonPrimitive(correlation.requestId))
            correlation.traceId?.let { put("traceId", JsonPrimitive(it)) }
            correlation.spanId?.let { put("spanId", JsonPrimitive(it)) }
            for ((key, value) in fields) {
                when (value) {
                    null -> put(key, JsonPrimitive("null"))
                    is Number -> put(key, JsonPrimitive(value))
                    is Boolean -> put(key, JsonPrimitive(value))
                    else -> put(key, JsonPrimitive(value.toString()))
                }
            }
        }
        println(payload.toString())
        System.out.flush()
    }
}

File: demos/09-full-platform/product/classify-python/jsonlog.py
"""Structured JSON logging to stdout (epic 01 required fields)."""

from __future__ import annotations

import json
import sys
from datetime import datetime, timezone
from typing import Any


_LEVEL_RANK = {"debug": 10, "info": 20, "warn": 30, "error": 40}


class Logger:
    def __init__(self, service: str, level: str) -> None:
        self._service = service
        self._min = _LEVEL_RANK.get(level.lower(), 20)

    def _emit(self, level: str, message: str, **fields: Any) -> None:
        if _LEVEL_RANK.get(level, 20) < self._min:
            return
        payload: dict[str, Any] = {
            "timestamp": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
            "level": level,
            "service": self._service,
            "message": message,
        }
        payload.update(fields)
        sys.stdout.write(json.dumps(payload, separators=(",", ":")) + "\n")
        sys.stdout.flush()

    def debug(self, message: str, **fields: Any) -> None:
        self._emit("debug", message, **fields)

    def info(self, message: str, **fields: Any) -> None:
        self._emit("info", message, **fields)

    def warn(self, message: str, **fields: Any) -> None:
        self._emit("warn", message, **fields)

    def error(self, message: str, **fields: Any) -> None:
        self._emit("error", message, **fields)
